src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs

```diff
@@ -146,17 +147,21 @@ protected virtual void MessageHandling((IActorRef, object) args)
 
         private void Pump()
         {
-            if (IsAvailable(_shape.Outlet))
+            Log.Debug("Pump pulled");
+            
+            while(IsAvailable(_shape.Outlet) && _buffer.TryDequeue(out var message))
```

Comment:
No need for recursion here - just run in a loop until the downstream can't accept any additional messages.